src/main/java/com/iota/iri/crypto/batched/BCTCurl.java

```diff
@@ -0,0 +1,121 @@
+package com.iota.iri.crypto.batched;
+
+/**
+ * A Curl implementation which absorbs binary-encoded-ternary inputs.
+ */
+public class BCTCurl {
+
+    private static final long HIGH_LONG_BITS = 0xFFFF_FFFF_FFFF_FFFFL;
+
+    private int hashLength;
+    private int numberOfRounds;
+    private int stateLength;
+    private BCTrinary state;
+
+    /**
+     * Creates a new {@link BCTCurl} with the given hash length and number of rounds.
+     * @param hashLength the desired hash length
+     * @param numberOfRounds the number of hashing rounds to apply
+     */
+    public BCTCurl(int hashLength, int numberOfRounds) {
+        this.hashLength = hashLength;
+        this.numberOfRounds = numberOfRounds;
+        this.stateLength = 3 * hashLength;
+        this.state = new BCTrinary(new long[3 * hashLength], new long[3 * hashLength]);
+        reset();
+    }
+
+    /**
+     * Resets the state of the hashing function.
+     */
+    public void reset() {
+        for (int i = 0; i < stateLength; i++) {
+            state.low[i] = HIGH_LONG_BITS;
+            state.high[i] = HIGH_LONG_BITS;
+        }
```

Comment:
This might improve performance and readability:

```java

public void reset() {
  Arrays.fill(state.low, HIGH_LONG_BITS);
  Arrays.fill(state.high, HIGH_LONG_BITS);
}